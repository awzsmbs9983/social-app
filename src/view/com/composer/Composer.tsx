import React, {useCallback, useEffect, useMemo, useRef, useState} from 'react'
import {observer} from 'mobx-react-lite'
import {
  ActivityIndicator,
  KeyboardAvoidingView,
  Platform,
  ScrollView,
  StyleSheet,
  TouchableOpacity,
  TouchableWithoutFeedback,
  View,
} from 'react-native'
import {useSafeAreaInsets} from 'react-native-safe-area-context'
import LinearGradient from 'react-native-linear-gradient'
import {FontAwesomeIcon} from '@fortawesome/react-native-fontawesome'
import {RichText} from '@atproto/api'
import {useAnalytics} from 'lib/analytics'
import {UserAutocompleteModel} from 'state/models/discovery/user-autocomplete'
import {ExternalEmbed} from './ExternalEmbed'
import {Text} from '../util/text/Text'
import * as Toast from '../util/Toast'
import {TextInput, TextInputRef} from './text-input/TextInput'
import {CharProgress} from './char-progress/CharProgress'
import {UserAvatar} from '../util/UserAvatar'
import {useStores} from 'state/index'
import * as apilib from 'lib/api/index'
import {ComposerOpts} from 'state/models/ui/shell'
import {s, colors, gradients} from 'lib/styles'
import {sanitizeDisplayName} from 'lib/strings/display-names'
import {cleanError} from 'lib/strings/errors'
import {SelectPhotoBtn} from './photos/SelectPhotoBtn'
import {OpenCameraBtn} from './photos/OpenCameraBtn'
import {usePalette} from 'lib/hooks/usePalette'
import QuoteEmbed from '../util/post-embeds/QuoteEmbed'
import {useExternalLinkFetch} from './useExternalLinkFetch'
import {isDesktopWeb, isAndroid} from 'platform/detection'
import {GalleryModel} from 'state/models/media/gallery'
import {Gallery} from './photos/Gallery'

const MAX_GRAPHEME_LENGTH = 300

type Props = ComposerOpts & {
  onClose: () => void
}

export const ComposePost = observer(function ComposePost({
  replyTo,
  onPost,
  onClose,
  quote: initQuote,
}: Props) {
  const {track} = useAnalytics()
  const pal = usePalette('default')
  const store = useStores()
  const textInput = useRef<TextInputRef>(null)
  const [isProcessing, setIsProcessing] = useState(false)
  const [processingState, setProcessingState] = useState('')
  const [error, setError] = useState('')
  const [richtext, setRichText] = useState(new RichText({text: ''}))
  const graphemeLength = useMemo(() => richtext.graphemeLength, [richtext])
  const [quote, setQuote] = useState<ComposerOpts['quote'] | undefined>(
    initQuote,
  )
  const {extLink, setExtLink} = useExternalLinkFetch({setQuote})
  const [suggestedLinks, setSuggestedLinks] = useState<Set<string>>(new Set())
  const gallery = useMemo(() => new GalleryModel(store), [store])

  const autocompleteView = useMemo<UserAutocompleteModel>(
    () => new UserAutocompleteModel(store),
    [store],
  )

  const insets = useSafeAreaInsets()

  // HACK
  // there's a bug with @mattermost/react-native-paste-input where if the input
  // is focused during unmount, an exception will throw (seems that a blur method isnt implemented)
  // manually blurring before closing gets around that
  // -prf
  const hackfixOnClose = useCallback(() => {
    textInput.current?.blur()
    onClose()
  }, [textInput, onClose])

  // initial setup
  useEffect(() => {
    autocompleteView.setup()
  }, [autocompleteView])

  useEffect(() => {
    // HACK
    // wait a moment before focusing the input to resolve some layout bugs with the keyboard-avoiding-view
    // -prf
    let to: NodeJS.Timeout | undefined
    if (textInput.current) {
      to = setTimeout(() => {
        textInput.current?.focus()
      }, 250)
    }
    return () => {
      if (to) {
        clearTimeout(to)
      }
    }
  }, [])

  const onPressContainer = useCallback(() => {
    textInput.current?.focus()
  }, [textInput])

  const onPressAddLinkCard = useCallback(
    (uri: string) => {
      setExtLink({uri, isLoading: true})
    },
    [setExtLink],
  )

  const onPhotoPasted = useCallback(
    async (uri: string) => {
      track('Composer:PastedPhotos')
      gallery.paste(uri)
    },
    [gallery, track],
  )

  const onPressPublish = useCallback(
    async rt => {
      if (isProcessing || rt.graphemeLength_ > MAX_GRAPHEME_LENGTH) {
        return
      }

      setError('')

      if (rt.text.trim().length === 0 && gallery.isEmpty) {
        setError('Did you want to say anything?')
        return false
      }

      setIsProcessing(true)

      let createdPost
      try {
        createdPost = await apilib.post(store, {
          rawText: rt.text,
          replyTo: replyTo?.uri,
          images: gallery.images,
          quote: quote,
          extLink: extLink,
          onStateChange: setProcessingState,
          knownHandles: autocompleteView.knownHandles,
        })
        track('Create Post', {
          imageCount: gallery.size,
        })
      } catch (e: any) {
        if (extLink) {
          setExtLink({
            ...extLink,
            isLoading: true,
            localThumb: undefined,
          } as apilib.ExternalEmbedDraft)
        }
        setError(cleanError(e.message))
        setIsProcessing(false)
        return
      }
      if (!replyTo) {
        store.me.mainFeed.addPostToTop(createdPost.uri)
      }
      onPost?.()
      hackfixOnClose()
      Toast.show(`Your ${replyTo ? 'reply' : 'post'} has been published`)
    },
    [
      isProcessing,
      setError,
      setIsProcessing,
      replyTo,
      autocompleteView.knownHandles,
      extLink,
      hackfixOnClose,
      onPost,
      quote,
      setExtLink,
      store,
      track,
      gallery,
    ],
  )

  const canPost = graphemeLength <= MAX_GRAPHEME_LENGTH

  const selectTextInputPlaceholder = replyTo ? 'Write your reply' : "What's up?"

  const canSelectImages = gallery.size < 4
  const viewStyles = {
    paddingBottom: isAndroid ? insets.bottom : 0,
    paddingTop: isAndroid ? insets.top : isDesktopWeb ? 0 : 15,
  }

  return (
    <KeyboardAvoidingView
      testID="composePostView"
      behavior={Platform.OS === 'ios' ? 'padding' : 'height'}
      style={styles.outer}>
      <TouchableWithoutFeedback onPressIn={onPressContainer}>
        <View style={[s.flex1, viewStyles]}>
          <View style={styles.topbar}>
            <TouchableOpacity
              testID="composerCancelButton"
              onPress={hackfixOnClose}>
              <Text style={[pal.link, s.f18]}>Cancel</Text>
            </TouchableOpacity>
            <View style={s.flex1} />
            {isProcessing ? (
              <View style={styles.postBtn}>
                <ActivityIndicator />
              </View>
            ) : canPost ? (
              <TouchableOpacity
                testID="composerPublishBtn"
                onPress={() => {
                  onPressPublish(richtext)
                }}>
                <LinearGradient
                  colors={[gradients.blueLight.start, gradients.blueLight.end]}
                  start={{x: 0, y: 0}}
                  end={{x: 1, y: 1}}
                  style={styles.postBtn}>
                  <Text style={[s.white, s.f16, s.bold]}>
                    {replyTo ? 'Reply' : 'Post'}
                  </Text>
                </LinearGradient>
              </TouchableOpacity>
            ) : (
              <View style={[styles.postBtn, pal.btn]}>
                <Text style={[pal.textLight, s.f16, s.bold]}>Post</Text>
              </View>
            )}
          </View>
          {isProcessing ? (
            <View style={[pal.btn, styles.processingLine]}>
              <Text style={pal.text}>{processingState}</Text>
            </View>
          ) : undefined}
          {error !== '' && (
            <View style={styles.errorLine}>
              <View style={styles.errorIcon}>
                <FontAwesomeIcon
                  icon="exclamation"
                  style={{color: colors.red4}}
                  size={10}
                />
              </View>
              <Text style={[s.red4, s.flex1]}>{error}</Text>
            </View>
          )}
          <ScrollView
            style={styles.scrollView}
            keyboardShouldPersistTaps="always">
            {replyTo ? (
              <View style={[pal.border, styles.replyToLayout]}>
                <UserAvatar avatar={replyTo.author.avatar} size={50} />
                <View style={styles.replyToPost}>
                  <Text type="xl-medium" style={[pal.text]}>
                    {sanitizeDisplayName(
                      replyTo.author.displayName || replyTo.author.handle,
                    )}
                  </Text>
                  <Text type="post-text" style={pal.text} numberOfLines={6}>
                    {replyTo.text}
                  </Text>
                </View>
              </View>
            ) : undefined}

            <View style={[pal.border, styles.textInputLayout]}>
              <UserAvatar avatar={store.me.avatar} size={50} />
              <TextInput
                ref={textInput}
                richtext={richtext}
                placeholder={selectTextInputPlaceholder}
                suggestedLinks={suggestedLinks}
                autocompleteView={autocompleteView}
                setRichText={setRichText}
                onPhotoPasted={onPhotoPasted}
                onPressPublish={onPressPublish}
                onSuggestedLinksChanged={setSuggestedLinks}
                onError={setError}
              />
            </View>

            <Gallery gallery={gallery} />
            {gallery.isEmpty && extLink && (
              <ExternalEmbed
                link={extLink}
                onRemove={() => setExtLink(undefined)}
              />
            )}
            {quote ? (
              <View style={s.mt5}>
                <QuoteEmbed quote={quote} />
              </View>
            ) : undefined}
          </ScrollView>
          {!extLink && suggestedLinks.size > 0 ? (
            <View style={s.mb5}>
              {Array.from(suggestedLinks).map(url => (
                <TouchableOpacity
                  key={`suggested-${url}`}
                  testID="addLinkCardBtn"
                  style={[pal.borderDark, styles.addExtLinkBtn]}
                  onPress={() => onPressAddLinkCard(url)}>
                  <Text style={pal.text}>
                    Add link card: <Text style={pal.link}>{url}</Text>
                  </Text>
                </TouchableOpacity>
              ))}
            </View>
          ) : null}
          <View style={[pal.border, styles.bottomBar]}>
            {canSelectImages ? (
              <>
                <SelectPhotoBtn gallery={gallery} />
                <OpenCameraBtn gallery={gallery} />
              </>
            ) : null}
            <View style={s.flex1} />
            <CharProgress count={graphemeLength} />
          </View>
        </View>
      </TouchableWithoutFeedback>
    </KeyboardAvoidingView>
  )
})

const styles = StyleSheet.create({
  outer: {
    flexDirection: 'column',
    flex: 1,
    height: '100%',
  },
  topbar: {
    flexDirection: 'row',
    alignItems: 'center',
    paddingTop: isDesktopWeb ? 10 : undefined,
    paddingBottom: 10,
    paddingHorizontal: 20,
    height: 55,
  },
  postBtn: {
    borderRadius: 20,
    paddingHorizontal: 20,
    paddingVertical: 6,
  },
  processingLine: {
    borderRadius: 6,
    paddingHorizontal: 8,
    paddingVertical: 6,
    marginHorizontal: 15,
    marginBottom: 6,
  },
  errorLine: {
    flexDirection: 'row',
    backgroundColor: colors.red1,
    borderRadius: 6,
    marginHorizontal: 15,
    paddingHorizontal: 8,
    paddingVertical: 6,
    marginVertical: 6,
  },
  errorIcon: {
    borderWidth: 1,
    borderColor: colors.red4,
    color: colors.red4,
    borderRadius: 30,
    width: 16,
    height: 16,
    alignItems: 'center',
    justifyContent: 'center',
    marginRight: 5,
  },
  scrollView: {
    flex: 1,
    paddingHorizontal: 15,
  },
  textInputLayout: {
    flex: isDesktopWeb ? undefined : 1,
    flexDirection: 'row',
    borderTopWidth: 1,
    paddingTop: 16,
  },
  replyToLayout: {
    flexDirection: 'row',
    borderTopWidth: 1,
    paddingTop: 16,
    paddingBottom: 16,
  },
  replyToPost: {
    flex: 1,
    paddingLeft: 13,
    paddingRight: 8,
  },
  addExtLinkBtn: {
    borderWidth: 1,
    borderRadius: 24,
    paddingHorizontal: 16,
    paddingVertical: 12,
    marginHorizontal: 10,
    marginBottom: 4,
  },
  bottomBar: {
    flexDirection: 'row',
    paddingVertical: 10,
    paddingLeft: 15,
    paddingRight: 20,
    alignItems: 'center',
    borderTopWidth: 1,
  },
})
