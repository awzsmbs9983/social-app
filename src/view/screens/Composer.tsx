import React, {useLayoutEffect, useRef} from 'react'
// import {Text, TouchableOpacity} from 'react-native'
// import {FontAwesomeIcon} from '@fortawesome/react-native-fontawesome'
import {Composer as ComposerComponent} from '../com/composer/Composer'
import {ScreenParams} from '../routes'

export const Composer = ({params}: ScreenParams) => {
  const {replyTo} = params
  const ref = useRef<{publish: () => Promise<boolean>}>()

  // TODO
  // useLayoutEffect(() => {
  //   navigation.setOptions({
  //     headerShown: true,
  //     headerTitle: replyTo ? 'Reply' : 'New Post',
  //     headerLeft: () => (
  //       <TouchableOpacity onPress={() => navigation.goBack()}>
  //         <FontAwesomeIcon icon="x" />
  //       </TouchableOpacity>
  //     ),
  //     headerRight: () => (
  //       <TouchableOpacity
  //         onPress={() => {
  //           if (!ref.current) {
  //             return
  //           }
  //           ref.current.publish().then(
  //             posted => {
  //               if (posted) {
  //                 navigation.goBack()
  //               }
  //             },
  //             err => console.error('Failed to create post', err),
  //           )
  //         }}>
  //         <Text>Post</Text>
  //       </TouchableOpacity>
  //     ),
  //   })
  // }, [navigation, replyTo, ref])

  return <ComposerComponent ref={ref} replyTo={replyTo} />
}
