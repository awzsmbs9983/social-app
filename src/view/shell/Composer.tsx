import React from 'react'
import {Modal, View} from 'react-native'
import {observer} from 'mobx-react-lite'

import {Provider as LegacyModalProvider} from '#/state/modals'
import {useComposerState} from 'state/shell/composer'
import {ModalsContainer as LegacyModalsContainer} from '#/view/com/modals/Modal'
import {useTheme} from '#/alf'
import {
  Outlet as PortalOutlet,
  Provider as PortalProvider,
} from '#/components/Portal'
import {ComposePost, useComposerCancelRef} from '../com/composer/Composer'

export const Composer = observer(function ComposerImpl({}: {
  winHeight: number
}) {
  const t = useTheme()
  const state = useComposerState()
  const ref = useComposerCancelRef()

  return (
    <Modal
      aria-modal
      accessibilityViewIsModal
      visible={!!state}
      presentationStyle="overFullScreen"
      animationType="slide"
      onRequestClose={() => ref.current?.onPressCancel()}>
      <View style={[t.atoms.bg, {flex: 1}]}>
        <LegacyModalProvider>
          <PortalProvider>
            <ComposePost
              cancelRef={ref}
              replyTo={state?.replyTo}
              onPost={state?.onPost}
              quote={state?.quote}
              mention={state?.mention}
              text={state?.text}
              imageUris={state?.imageUris}
            />
            <LegacyModalsContainer />
            <PortalOutlet />
          </PortalProvider>
        </LegacyModalProvider>
      </View>
    </Modal>
  )
})
