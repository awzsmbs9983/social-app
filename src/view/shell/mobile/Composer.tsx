import React, {useEffect} from 'react'
import {observer} from 'mobx-react-lite'
import {Animated, Easing, Platform, StyleSheet, View} from 'react-native'
import {ComposePost} from '../../com/composer/ComposePost'
import {ComposerOpts} from '../../../state/models/shell-ui'
import {useAnimatedValue} from '../../lib/hooks/useAnimatedValue'

export const Composer = observer(
  ({
    active,
    winHeight,
    replyTo,
    onPost,
    onClose,
  }: {
    active: boolean
    winHeight: number
    replyTo?: ComposerOpts['replyTo']
    onPost?: ComposerOpts['onPost']
    onClose: () => void
  }) => {
    const initInterp = useAnimatedValue(0)

    useEffect(() => {
      if (active) {
        Animated.timing(initInterp, {
          toValue: 1,
          duration: 300,
          easing: Easing.out(Easing.exp),
          useNativeDriver: true,
        }).start()
      } else {
        initInterp.setValue(0)
      }
    }, [initInterp, active])
    const wrapperAnimStyle = {
      transform: [
        {
          translateY: initInterp.interpolate({
            inputRange: [0, 1],
            outputRange: [winHeight, 0],
          }),
        },
      ],
    }

    // events
    // =

    // rendering
    // =

    if (!active) {
      return <View />
    }

    return (
      <Animated.View style={[styles.wrapper, wrapperAnimStyle]}>
        <ComposePost replyTo={replyTo} onPost={onPost} onClose={onClose} />
      </Animated.View>
    )
  },
)

const styles = StyleSheet.create({
  wrapper: {
    position: 'absolute',
    top: 0,
    bottom: 0,
    width: '100%',
    backgroundColor: '#fff',
    ...Platform.select({
      ios: {
        paddingTop: 24,
      },
    }),
  },
})
