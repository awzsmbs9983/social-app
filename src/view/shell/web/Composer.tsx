import React from 'react'
import {observer} from 'mobx-react-lite'
import {StyleSheet, View} from 'react-native'
import {ComposePost} from '../../com/composer/ComposePost'
import {ComposerOpts} from 'state/models/shell-ui'
import {usePalette} from 'lib/hooks/usePalette'

export const Composer = observer(
  ({
    active,
    replyTo,
    imagesOpen,
    onPost,
    onClose,
  }: {
    active: boolean
    winHeight: number
    replyTo?: ComposerOpts['replyTo']
    imagesOpen?: ComposerOpts['imagesOpen']
    onPost?: ComposerOpts['onPost']
    onClose: () => void
  }) => {
    const pal = usePalette('default')

    // rendering
    // =

    if (!active) {
      return <View />
    }

    return (
      <View style={styles.mask}>
        <View style={[styles.container, pal.view]}>
          <ComposePost
            replyTo={replyTo}
            imagesOpen={imagesOpen}
            onPost={onPost}
            onClose={onClose}
          />
        </View>
      </View>
    )
  },
)

const styles = StyleSheet.create({
  mask: {
    position: 'absolute',
    top: 0,
    left: 0,
    width: '100%',
    height: '100%',
    backgroundColor: '#000c',
    alignItems: 'center',
    justifyContent: 'center',
  },
  container: {
    maxWidth: 600,
    width: '100%',
    paddingVertical: 0,
    paddingHorizontal: 2,
    borderRadius: 8,
    marginBottom: '10vh',
  },
})
