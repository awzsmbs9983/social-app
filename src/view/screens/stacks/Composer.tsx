import React, {useLayoutEffect, useRef} from 'react'
import {Text, TouchableOpacity} from 'react-native'
import {FontAwesomeIcon} from '@fortawesome/react-native-fontawesome'
import {Shell} from '../../shell'
import type {RootTabsScreenProps} from '../../routes/types'
import {Composer as ComposerComponent} from '../../com/composer/Composer'

export const Composer = ({
  navigation,
  route,
}: RootTabsScreenProps<'Composer'>) => {
  const {replyTo} = route.params
  const ref = useRef<{publish: () => Promise<boolean>}>()

  useLayoutEffect(() => {
    navigation.setOptions({
      headerShown: true,
      headerTitle: replyTo ? 'Reply' : 'New Post',
      headerLeft: () => (
        <TouchableOpacity onPress={() => navigation.goBack()}>
          <FontAwesomeIcon icon="x" />
        </TouchableOpacity>
      ),
      headerRight: () => (
        <TouchableOpacity
          onPress={() => {
            if (!ref.current) {
              return
            }
            ref.current.publish().then(
              posted => {
                if (posted) {
                  navigation.goBack()
                }
              },
              err => console.error('Failed to create post', err),
            )
          }}>
          <Text>Post</Text>
        </TouchableOpacity>
      ),
    })
  }, [navigation, replyTo, ref])

  return (
    <Shell>
      <ComposerComponent ref={ref} replyTo={replyTo} />
    </Shell>
  )
}
