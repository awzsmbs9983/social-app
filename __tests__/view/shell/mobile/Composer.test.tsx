import React from 'react'
import {Composer} from '../../../../src/view/shell/mobile/Composer'
import renderer from 'react-test-renderer'
// import {render} from '../../../../jest/test-utils'

describe('Composer', () => {
  const mockedProps = {
    active: true,
    winHeight: 844,
    replyTo: {
      author: {avatar: undefined, displayName: 'Alice', handle: 'alice.test'},
      cid: 'bafyreieucrv36ylxrut4dr4jj264q2jj2vt2vfvhjfchgw3vua4gksvzia',
      text: 'Captain, maybe we ought to turn on the searchlights now. No… that’s just what they’ll be expecting us to do.',
      uri: 'at://did:plc:v3xz273ea2dzjpu2szsjzfue/app.bsky.feed.post/3jkcir3fhqv2u',
    },
    onPost: jest.fn(),
    onClose: jest.fn(),
  }
  it('renders correctly', () => {
    const tree = renderer.create(<Composer {...mockedProps} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
