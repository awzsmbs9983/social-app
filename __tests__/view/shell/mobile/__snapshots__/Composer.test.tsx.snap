// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Composer renders correctly 1`] = `
<View
  collapsable={false}
  style={
    Object {
      "backgroundColor": "#fff",
      "bottom": 0,
      "paddingTop": 24,
      "position": "absolute",
      "top": 0,
      "transform": Array [
        Object {
          "translateY": 844,
        },
      ],
      "width": "100%",
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "flex": 1,
          "flexDirection": "column",
          "height": "100%",
          "padding": 15,
          "paddingBottom": 0,
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <RCTSafeAreaView
      emulateUnlessSupported={true}
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 55,
            "paddingBottom": 10,
            "paddingHorizontal": 5,
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                },
                Array [
                  Object {
                    "color": "#0085ff",
                  },
                  Object {
                    "fontSize": 18,
                  },
                ],
              ]
            }
          >
            Cancel
          </Text>
        </View>
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        />
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <BVLinearGradient
            colors={
              Array [
                4292542719,
                4294901882,
              ]
            }
            endPoint={
              Object {
                "x": 1,
                "y": 1,
              }
            }
            locations={null}
            startPoint={
              Object {
                "x": 0,
                "y": 0,
              }
            }
            style={
              Object {
                "borderRadius": 20,
                "paddingHorizontal": 20,
                "paddingVertical": 6,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000000",
                  },
                  Array [
                    Object {
                      "color": "#ffffff",
                    },
                    Object {
                      "fontSize": 16,
                    },
                    Object {
                      "fontWeight": "bold",
                    },
                  ],
                ]
              }
            >
              Reply
            </Text>
          </BVLinearGradient>
        </View>
      </View>
      <RCTScrollView
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <View>
          <View
            style={
              Object {
                "borderTopColor": "#e4e2e2",
                "borderTopWidth": 1,
                "flexDirection": "row",
                "paddingBottom": 16,
                "paddingTop": 16,
              }
            }
          >
            <RNSVGSvgView
              align="xMidYMid"
              bbHeight={50}
              bbWidth={50}
              focusable={false}
              height={50}
              meetOrSlice={0}
              minX={0}
              minY={0}
              style={
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "borderWidth": 0,
                  },
                  Object {
                    "flex": 0,
                    "height": 50,
                    "width": 50,
                  },
                ]
              }
              vbHeight={100}
              vbWidth={100}
              width={50}
            >
              <RNSVGGroup>
                <RNSVGDefs>
                  <RNSVGLinearGradient
                    gradient={
                      Array [
                        0,
                        -16742913,
                        1,
                        -14631929,
                      ]
                    }
                    gradientTransform={null}
                    gradientUnits={0}
                    name="grad"
                    x1="0"
                    x2="1"
                    y1="0"
                    y2="1"
                  />
                </RNSVGDefs>
                <RNSVGCircle
                  cx="50"
                  cy="50"
                  fill={
                    Array [
                      1,
                      "grad",
                    ]
                  }
                  propList={
                    Array [
                      "fill",
                    ]
                  }
                  r="50"
                />
                <RNSVGText
                  content={null}
                  dx={Array []}
                  dy={Array []}
                  fill={4294967295}
                  font={
                    Object {
                      "fontSize": "50",
                      "fontWeight": "bold",
                      "textAnchor": "middle",
                    }
                  }
                  propList={
                    Array [
                      "fill",
                    ]
                  }
                  rotate={Array []}
                  x={
                    Array [
                      "50",
                    ]
                  }
                  y={
                    Array [
                      "67",
                    ]
                  }
                >
                  <RNSVGTSpan
                    content="A"
                    dx={Array []}
                    dy={Array []}
                    font={Object {}}
                    rotate={Array []}
                    x={Array []}
                    y={Array []}
                  />
                </RNSVGText>
              </RNSVGGroup>
            </RNSVGSvgView>
            <View
              style={
                Object {
                  "flex": 1,
                  "paddingLeft": 13,
                  "paddingRight": 8,
                }
              }
            >
              <Text
                onLongPress={[Function]}
                onPress={[Function]}
                style={
                  Array [
                    Object {
                      "color": "#000000",
                    },
                    Array [
                      Object {
                        "fontSize": 16,
                      },
                      Object {
                        "fontWeight": "bold",
                      },
                      Object {
                        "color": "#000000",
                      },
                    ],
                  ]
                }
              >
                Alice
              </Text>
              <Text
                numberOfLines={6}
                style={
                  Array [
                    Object {
                      "color": "#000000",
                    },
                    Array [
                      Object {
                        "fontSize": 16,
                      },
                      Object {
                        "lineHeight": 20.8,
                      },
                      Object {
                        "color": "#000000",
                      },
                    ],
                  ]
                }
              >
                Captain, maybe we ought to turn on the searchlights now. No… that’s just what they’ll be expecting us to do.
              </Text>
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "borderTopColor": "#e4e2e2",
                  "borderTopWidth": 1,
                  "flexDirection": "row",
                  "paddingTop": 16,
                },
                Object {
                  "flex": 1,
                },
              ]
            }
          >
            <RNSVGSvgView
              align="xMidYMid"
              bbHeight={50}
              bbWidth={50}
              focusable={false}
              height={50}
              meetOrSlice={0}
              minX={0}
              minY={0}
              style={
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "borderWidth": 0,
                  },
                  Object {
                    "flex": 0,
                    "height": 50,
                    "width": 50,
                  },
                ]
              }
              vbHeight={100}
              vbWidth={100}
              width={50}
            >
              <RNSVGGroup>
                <RNSVGDefs>
                  <RNSVGLinearGradient
                    gradient={
                      Array [
                        0,
                        -1292135,
                        1,
                        -2424577,
                      ]
                    }
                    gradientTransform={null}
                    gradientUnits={0}
                    name="grad"
                    x1="0"
                    x2="1"
                    y1="0"
                    y2="1"
                  />
                </RNSVGDefs>
                <RNSVGCircle
                  cx="50"
                  cy="50"
                  fill={
                    Array [
                      1,
                      "grad",
                    ]
                  }
                  propList={
                    Array [
                      "fill",
                    ]
                  }
                  r="50"
                />
                <RNSVGText
                  content={null}
                  dx={Array []}
                  dy={Array []}
                  fill={4294967295}
                  font={
                    Object {
                      "fontSize": "50",
                      "fontWeight": "bold",
                      "textAnchor": "middle",
                    }
                  }
                  propList={
                    Array [
                      "fill",
                    ]
                  }
                  rotate={Array []}
                  x={
                    Array [
                      "50",
                    ]
                  }
                  y={
                    Array [
                      "67",
                    ]
                  }
                >
                  <RNSVGTSpan
                    content="X"
                    dx={Array []}
                    dy={Array []}
                    font={Object {}}
                    rotate={Array []}
                    x={Array []}
                    y={Array []}
                  />
                </RNSVGText>
              </RNSVGGroup>
            </RNSVGSvgView>
            <TextInput
              multiline={true}
              onChangeText={[Function]}
              placeholder="Write your reply"
              placeholderTextColor="#968d8d"
              scrollEnabled={true}
              style={
                Object {
                  "alignSelf": "flex-start",
                  "color": "#000000",
                  "flex": 1,
                  "fontSize": 18,
                  "marginLeft": 8,
                  "padding": 5,
                }
              }
            />
          </View>
        </View>
      </RCTScrollView>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "borderTopColor": "#e4e2e2",
            "borderTopWidth": 1,
            "flexDirection": "row",
            "paddingRight": 5,
            "paddingVertical": 10,
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          hitSlop={
            Object {
              "bottom": 10,
              "left": 10,
              "right": 10,
              "top": 10,
            }
          }
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
              "paddingLeft": 5,
            }
          }
        >
          <
            icon={
              Array [
                "far",
                "image",
              ]
            }
            size={24}
            style={
              Object {
                "color": "#0085ff",
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        />
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
              },
              Array [
                Object {
                  "marginRight": 10,
                },
                Object {
                  "color": undefined,
                },
              ],
            ]
          }
        >
          256
        </Text>
        <View>
          <View
            indeterminateAnimationDuration={1000}
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "overflow": "hidden",
                },
                undefined,
              ]
            }
          >
            <RNSVGSvgView
              bbHeight={30}
              bbWidth={30}
              collapsable={false}
              focusable={false}
              height={30}
              style={
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "borderWidth": 0,
                  },
                  Object {},
                  Object {
                    "flex": 0,
                    "height": 30,
                    "width": 30,
                  },
                ]
              }
              width={30}
            >
              <RNSVGGroup>
                <RNSVGPath
                  d="M15 2.5
          A12.5 12.5 0 0 1 15 2.5"
                  propList={
                    Array [
                      "stroke",
                      "strokeWidth",
                      "strokeLinecap",
                    ]
                  }
                  stroke={4278221567}
                  strokeLinecap={0}
                  strokeWidth={3}
                />
                <RNSVGPath
                  d="M15 0.5
            a14.5 14.5 0 0 1 0 29
            a14.5 14.5 0 0 1 0 -29"
                  propList={
                    Array [
                      "stroke",
                      "strokeWidth",
                      "strokeLinecap",
                    ]
                  }
                  stroke={4293190370}
                  strokeLinecap={0}
                  strokeWidth={1}
                />
              </RNSVGGroup>
            </RNSVGSvgView>
          </View>
        </View>
      </View>
      <View
        collapsable={false}
        style={
          Object {
            "backgroundColor": "#ffffff",
            "borderTopColor": "#e4e2e2",
            "borderTopWidth": 1,
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 1334,
          }
        }
      />
    </RCTSafeAreaView>
  </View>
</View>
`;
